code/datums/elements/tool_renaming.dm

```diff
@@ -0,0 +1,67 @@
+/**
+ * Renaming tool element
+ *
+ * When using this tool on an object with UNIQUE_RENAME,
+ * lets the user rename/redesc it.
+ */
+/datum/element/tool_renaming
+
+/datum/element/tool_renaming/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ITEM_INTERACTING_WITH_ATOM, PROC_REF(attempt_rename))
+
+/datum/element/tool_renaming/Detach(datum/source)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ITEM_INTERACTING_WITH_ATOM)
+
+/datum/element/tool_renaming/proc/attempt_rename(datum/source, mob/living/user, atom/interacting_with, list/modifiers)
+	SIGNAL_HANDLER
+
+	if(!isobj(interacting_with))
+		return NONE
+	
+	var/obj/renamed_obj = interacting_with
+
+	if(!(renamed_obj.obj_flags & UNIQUE_RENAME))
+		return NONE
+
+	var/pen_choice = tgui_input_list(user, "What would you like to edit?", "Pen Setting", list("Rename", "Description", "Reset"))
+	if(QDELETED(renamed_obj) || !user.can_perform_action(renamed_obj))
+		return ITEM_INTERACT_BLOCKING
+
+	switch(pen_choice)
+		if("Rename")
+			var/input = tgui_input_text(user, "What do you want to name [renamed_obj]?", "Object Name", "[renamed_obj.name]", MAX_NAME_LEN)
+			var/old_name = renamed_obj.name
+			if(QDELETED(renamed_obj) || !user.can_perform_action(renamed_obj))
+				return ITEM_INTERACT_BLOCKING
+			if(input == old_name || !input)
+				to_chat(user, span_notice("You changed [renamed_obj] to... well... [renamed_obj]."))
+				return ITEM_INTERACT_BLOCKING
+			renamed_obj.AddComponent(/datum/component/rename, input, renamed_obj.desc)
+			to_chat(user, span_notice("You have successfully renamed \the [old_name] to [renamed_obj]."))
+			ADD_TRAIT(renamed_obj, TRAIT_WAS_RENAMED, PEN_LABEL_TRAIT)
+			renamed_obj.update_appearance(UPDATE_ICON)
```

Comment:
```suggestion
			renamed_obj.update_appearance(UPDATE_NAME)
```
Surely?

Reply:
Looking into it, bodybags actually have their own custom implementation of this, for whichever reason.
https://github.com/tgstation/tgstation/blob/6e36ed984070d53e16bed4fa43eb0cbe6460deed/code/game/objects/structures/crates_lockers/closets/bodybag.dm#L49-L59

Reply:
But yes, I didn't actually consider this, I feel like *ideally* these should all use the same base logic though.
Either way, lemme look into every item using `UNIQUE_RENAME` to get an idea.

Reply:
So looking into this, I could find a few more things:
1. The fermenting barrel uses `UNIQUE_RENAME` and requires an icon update to add a label... but it also just registers a signal like Melbert mentioned, so I think this one's a non-issue.
2. I know food can be renamed, but if there's any foods that change their icon based on their name then I sure as hell wouldn't be able to find them in the massive pile of food files we have.
3. If there's any others, they probably also either use their own special renaming implementation or do it in a different file (inheritors/components/such).

I _think_ we're safe to change it and use the signal method going forward, but there's too many files I'd still need to go through to say I'm _completely_ sure.